File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos31/Alg31.java
/*
♦ Algoritmo 31: Algoritmo de implementação da Estrutura de Dados Não Linear Estática (não Dinâmica) de Grafo: Prim.
*/


import java.util.*;

// Classe para representar um vértice no grafo
class Vertice<T> implements Comparable<Vertice<T>> {
    int indice;
    T valor;

    public Vertice(int indice, T valor) {
        this.indice = indice;
        this.valor = valor;
    }

    // Método para comparar vértices com base em seus valores
    @Override
    public int compareTo(Vertice<T> outro) {
        if (this.valor instanceof Comparable) {
            return ((Comparable<T>) this.valor).compareTo(outro.valor);
        } else {
            throw new UnsupportedOperationException("Tipo de valor não suportado para comparação.");
        }
    }
}

// Classe para representar uma aresta no grafo ponderado
class Aresta<T> {
    Vertice<T> destino;
    double peso;

    public Aresta(Vertice<T> destino, double peso) {
        this.destino = destino;
        this.peso = peso;
    }
}

// Classe para representar um grafo ponderado usando lista de adjacência
class Grafo<T> {
    Map<Vertice<T>, List<Aresta<T>>> adjacencia;

    public Grafo() {
        adjacencia = new HashMap<>();
    }

    // Adiciona uma aresta ao grafo
    public void addAresta(Vertice<T> origem, Vertice<T> destino, double peso) {
        adjacencia.computeIfAbsent(origem, k -> new ArrayList<>()).add(new Aresta<>(destino, peso));
        adjacencia.computeIfAbsent(destino, k -> new ArrayList<>()).add(new Aresta<>(origem, peso)); // Para grafos não direcionados
    }

    // Algoritmo de Prim para encontrar a árvore geradora mínima
    public List<Aresta<T>> prim(Vertice<T> origem) {
        Set<Vertice<T>> visitados = new HashSet<>();
        List<Aresta<T>> arvoreGeradoraMinima = new ArrayList<>();
        PriorityQueue<Aresta<T>> filaPrioridade = new PriorityQueue<>(Comparator.comparingDouble(a -> a.peso));

        visitados.add(origem);
        adicionarArestasAdjacentes(origem, filaPrioridade, visitados);

        while (!filaPrioridade.isEmpty()) {
            Aresta<T> aresta = filaPrioridade.poll();
            Vertice<T> destino = aresta.destino;

            if (!visitados.contains(destino)) {
                visitados.add(destino);
                arvoreGeradoraMinima.add(aresta);
                adicionarArestasAdjacentes(destino, filaPrioridade, visitados);
            }
        }

        return arvoreGeradoraMinima;
    }

    private void adicionarArestasAdjacentes(Vertice<T> vertice, PriorityQueue<Aresta<T>> filaPrioridade, Set<Vertice<T>> visitados) {
        for (Aresta<T> aresta : adjacencia.getOrDefault(vertice, Collections.emptyList())) {
            if (!visitados.contains(aresta.destino)) {
                filaPrioridade.add(aresta);
            }
        }
    }
}

// Classe principal
public class Alg31 {
    public static void main(String[] args) {
        // Criando um grafo
        Grafo<String> grafo = new Grafo<>();

        // Criando vértices
        Vertice<String> A = new Vertice<>(0, "A");
        Vertice<String> B = new Vertice<>(1, "B");
        Vertice<String> C = new Vertice<>(2, "C");
        Vertice<String> D = new Vertice<>(3, "D");
        Vertice<String> E = new Vertice<>(4, "E");

        // Adicionando arestas ao grafo
        grafo.addAresta(A, B, 4);
        grafo.addAresta(A, C, 1);
        grafo.addAresta(B, C, 2);
        grafo.addAresta(B, D, 5);
        grafo.addAresta(C, D, 8);
        grafo.addAresta(C, E, 10);
        grafo.addAresta(D, E, 2);

        // Executando o algoritmo de Prim a partir do vértice A
        List<Aresta<String>> arvoreGeradoraMinima = grafo.prim(A);

        // Imprimindo a árvore geradora mínima
        for (Aresta<String> aresta : arvoreGeradoraMinima) {
            System.out.println("Aresta: " + aresta.destino.valor + ", Peso: " + aresta.peso);
        }
    }
}

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos31/Alg31.js
/*
♦ Algoritmo 31: Algoritmo de implementação da Estrutura de Dados Não Linear Estática (não Dinâmica) de Grafo: Prim.
*/


// Classe para representar um vértice no grafo
class Vertice {
    constructor(indice, valor) {
        this.indice = indice;
        this.valor = valor;
    }

    // Método para comparar vértices com base em seus valores
    compareTo(outro) {
        if (this.valor instanceof Comparable) {
            return this.valor.compareTo(outro.valor);
        } else {
            throw new Error("Tipo de valor não suportado para comparação.");
        }
    }
}

// Classe para representar uma aresta no grafo ponderado
class Aresta {
    constructor(destino, peso) {
        this.destino = destino;
        this.peso = peso;
    }
}

// Classe para representar um grafo ponderado usando lista de adjacência
class Grafo {
    constructor() {
        this.adjacencia = new Map();
    }

    // Adiciona uma aresta ao grafo
    addAresta(origem, destino, peso) {
        this.adjacencia.set(origem, this.adjacencia.get(origem) || []);
        this.adjacencia.set(destino, this.adjacencia.get(destino) || []);
        this.adjacencia.get(origem).push(new Aresta(destino, peso));
        this.adjacencia.get(destino).push(new Aresta(origem, peso)); // Para grafos não direcionados
    }

    // Algoritmo de Prim para encontrar a árvore geradora mínima
    prim(origem) {
        const visitados = new Set();
        const arvoreGeradoraMinima = [];
        const filaPrioridade = new PriorityQueue((a, b) => a.peso - b.peso);

        visitados.add(origem);
        this.adicionarArestasAdjacentes(origem, filaPrioridade, visitados);

        while (!filaPrioridade.isEmpty()) {
            const aresta = filaPrioridade.poll();
            const destino = aresta.destino;

            if (!visitados.has(destino)) {
                visitados.add(destino);
                arvoreGeradoraMinima.push(aresta);
                this.adicionarArestasAdjacentes(destino, filaPrioridade, visitados);
            }
        }

        return arvoreGeradoraMinima;
    }

    adicionarArestasAdjacentes(vertice, filaPrioridade, visitados) {
        for (const aresta of this.adjacencia.get(vertice) || []) {
            if (!visitados.has(aresta.destino)) {
                filaPrioridade.add(aresta);
            }
        }
    }
}

// Implementação de uma fila de prioridade
class PriorityQueue {
    constructor(comparator) {
        this.comparator = comparator;
        this.elements = [];
    }

    add(element) {
        this.elements.push(element);
        this.elements.sort(this.comparator);
    }

    poll() {
        return this.elements.shift();
    }

    isEmpty() {
        return this.elements.length === 0;
    }
}

// Classe principal
function main() {
    // Criando um grafo
    const grafo = new Grafo();

    // Criando vértices
    const A = new Vertice(0, "A");
    const B = new Vertice(1, "B");
    const C = new Vertice(2, "C");
    const D = new Vertice(3, "D");
    const E = new Vertice(4, "E");

    // Adicionando arestas ao grafo
    grafo.addAresta(A, B, 4);
    grafo.addAresta(A, C, 1);
    grafo.addAresta(B, C, 2);
    grafo.addAresta(B, D, 5);
    grafo.addAresta(C, D, 8);
    grafo.addAresta(C, E, 10);
    grafo.addAresta(D, E, 2);

    // Executando o algoritmo de Prim a partir do vértice A
    const arvoreGeradoraMinima = grafo.prim(A);

    // Imprimindo a árvore geradora mínima
    for (const aresta of arvoreGeradoraMinima) {
        console.log("Aresta: " + aresta.destino.valor + ", Peso: " + aresta.peso);
    }
}

// Executando a função principal
main();
